### PR TITLE
[SPARK-19762][ML] Hierarchy for consolidating ML aggregator/loss code

### PR DESCRIPTION
**Pull request by sethah**
## What changes were proposed in this pull request?

JIRA: [SPARK-19762](https://issues.apache.org/jira/browse/SPARK-19762)

The larger changes in this patch are:

* Adds a `DifferentiableLossAggregator` trait which is intended to be used as a common parent trait to all Spark ML aggregator classes. It factors out the common methods: `merge, gradient, loss, weight` from the aggregator subclasses.
* Adds a `RDDLossFunction` which is intended to be the only implementation of Breeze's `DiffFunction` necessary in Spark ML, and can be used by all other algorithms. It takes the aggregator type as a type parameter, and maps the aggregator over an RDD. It additionally takes in a optional regularization loss function for applying the differentiable part of regularization.
* Factors out the regularization from the data part of the cost function, and treats regularization as a separate independent cost function which can be evaluated and added to the data cost function.
* Changes `LinearRegression` to use this new hierarchy as a proof of concept.
* Adds the following new namespaces `o.a.s.ml.optim.loss` and `o.a.s.ml.optim.aggregator`

Also note that none of these are public-facing changes. All of these classes are internal to Spark ML and remain that way.

**NOTE: The large majority of the "lines added" and "lines deleted" are simply code moving around or unit tests.**

BTW, I also converted LinearSVC to this framework as a way to prove that this new hierarchy is flexible enough for the other algorithms, but I backed those changes out because the PR is large enough as is. 

## How was this patch tested?
Test suites are added for the new components, and some test suites are also added to provide coverage where there wasn't any before.

* DifferentiablLossAggregatorSuite
* LeastSquaresAggregatorSuite
* RDDLossFunctionSuite
* DifferentiableRegularizationSuite

Below are some performance testing numbers. Run on a 6 node virtual cluster with 44 cores and ~110G RAM, the dataset size is about 37G. These are not "large-scale" tests, but we really want to just make sure the iteration times don't increase with this patch. Notably we are doing the regularization a bit differently than before, but that should cost very little. I think there's very little risk otherwise, and these numbers don't show a difference. Of course I'm happy to add more tests as we think it's necessary, but I think the patch is ready for review now.

**Note:** timings are best of 3 runs. 

|    |   numFeatures |   numPoints |   maxIter |   regParam |   elasticNetParam |   SPARK-19762 (sec) |   master (sec) |
|----|---------------|-------------|-----------|------------|-------------------|---------------------|----------------|
|  0 |          5000 |       1e+06 |        30 |       0    |               0   |             129.594 |        131.153 |
|  1 |          5000 |       1e+06 |        30 |       0.1  |               0   |             135.54  |        136.327 |
|  2 |          5000 |       1e+06 |        30 |       0.01 |               0.5 |             135.148 |        129.771 |
|  3 |         50000 |  100000     |        30 |       0    |               0   |             145.764 |        144.096 |

## Follow ups

If this design is accepted, we will convert the other ML algorithms that use this aggregator pattern to this new hierarchy in follow up PRs.

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    Instance(1.0, 0.5, Vectors.dense(1.5, 1.0)),
+    Instance(2.0, 0.3, Vectors.dense(4.0, 0.5))
+  )
+  private val instances2 = Seq(
+    Instance(0.2, 0.4, Vectors.dense(0.8, 2.5)),
+    Instance(0.8, 0.9, Vectors.dense(2.0, 1.3)),
+    Instance(1.5, 0.2, Vectors.dense(3.0, 0.2))
+  )
+
+  def assertEqual[T, Agg <: DifferentiableLossAggregator[T, Agg]](
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala]: make private

File: mllib/src/main/scala/org/apache/spark/ml/optim/loss/RDDLossFunction.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ */
+private[ml] class RDDLossFunction[Agg <: DifferentiableLossAggregator[Instance, Agg]: ClassTag](
+    instances: RDD[Instance],
+    getAggregator: (Broadcast[Vector] => Agg),
+    regularization: Option[DifferentiableRegularization[Array[Double]]],
+    aggregationDepth: Int = 2)
+  extends DiffFunction[BDV[Double]] {
+
+  override def calculate(coefficients: BDV[Double]): (Double, BDV[Double]) = {
+    val bcCoefficients = instances.context.broadcast(Vectors.dense(coefficients.data))
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/optim/loss/RDDLossFunction.scala]: use `asBreeze`

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    assert(agg.weight ~== weightSum relTol 1e-5)
+    assert(agg.loss ~== expectedLoss.sum / weightSum relTol 1e-5)
+    assert(agg.gradient ~== expectedGradient relTol 1e-5)
+  }
+}
+
+/**
+ * Dummy aggregator that represents least squares cost with no intercept.
+ */
+class TestAggregator(numFeatures: Int)(coefficients: Vector)
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala]: move it into a companion object

**Comment by sethah** [comment]
ping @MLnick @jkbradley

**Comment by sethah** [comment]
Jenkins test this please.

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregator.scala [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  /** Array of gradient values that are mutated when new instances are added to the aggregator. */
+  protected lazy val gradientSumArray: Array[Double] = Array.ofDim[Double](dim)
+
+  /** Add a single data point to this aggregator. */
+  def add(instance: Datum): Agg
+
+  /** Merge two aggregators. The `this` object will be modified in place and returned. */
+  def merge(other: Agg): Agg = {
+    require(dim == other.dim, s"Dimensions mismatch when merging with another " +
+      s"LeastSquaresAggregator. Expecting $dim but got ${other.dim}.")
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregator.scala]: change this to use `getClass.getName`

**Comment by sethah** [comment]
Removed WIP, think it's ready now :)

File: mllib/src/main/scala/org/apache/spark/ml/optim/loss/DifferentiableRegularization.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+import breeze.optimize.DiffFunction
+
+/**
+ * A Breeze diff function which represents a cost function for differentiable regularization
+ * of parameters. e.g. L2 regularization: 1 / 2 regParam * <\beta, \beta>
+ *
+ * @tparam T The type of the coefficients being regularized.
+ */
+trait DifferentiableRegularization[T] extends DiffFunction[T] {
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/optim/loss/DifferentiableRegularization.scala]: make these private

**Comment by sethah** [comment]
Jenkins test this please.

**Comment by sethah** [comment]
ping! @MLnick @jkbradley @yanboliang @hhbyyh 

Is there any interest in this? I actually think this cleanup will be a precursor to several different improvements (adding more optimized aggregators, adding optimization library) and that it will be very useful. IMO it's an important change. Otherwise, we keep slapping layers on the current implementation and the code length and complexity keeps growing.

I'm happy to take suggestions, make changes, or discuss it further. Thoughts?

**Comment by MLnick** [comment]
In terms of the high level intention of this, agree we definitely need it and it should clean things up substantially. I will start taking a look through ASAP. Thanks!

**Comment by sethah** [comment]
Thanks @MLnick! I am happy to discuss splitting this into smaller bits as well, if it can make things easier.

**Comment by sethah** [comment]
cc @srowen also

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LeastSquaresAggregator.scala [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+ */
+private[ml] class LeastSquaresAggregator(
+    labelStd: Double,
+    labelMean: Double,
+    fitIntercept: Boolean,
+    bcFeaturesStd: Broadcast[Array[Double]],
+    bcFeaturesMean: Broadcast[Array[Double]])(bcCoefficients: Broadcast[Vector])
+  extends DifferentiableLossAggregator[Instance, LeastSquaresAggregator] {
+  require(labelStd > 0.0, s"${this.getClass.getName} requires the label standard" +
+    s"deviation to be positive.")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LeastSquaresAggregator.scala]: Add a space before 'deviation' or at the end of the previous line

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregator.scala [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @tparam Datum The type of the instances added to the aggregator to update the loss and gradient.
+ * @tparam Agg Specialization of [[DifferentiableLossAggregator]]. Classes that subclass this
+ *             type need to use this parameter to specify the concrete type of the aggregator.
+ */
+private[ml] trait DifferentiableLossAggregator[
+    Datum,
+    Agg <: DifferentiableLossAggregator[Datum, Agg]] extends Serializable {
+
+  self: Agg =>
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregator.scala]: You could add a brief comment explaining what this self type does

File: mllib/src/main/scala/org/apache/spark/ml/optim/loss/DifferentiableRegularization.scala [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        case Some(std) =>
+          if (std(j) != 0.0) {
+            val temp = coefficients(j) / (std(j) * std(j))
+            sum += coefficients(j) * temp
+            gradient(j) = regParam * temp
+          } else {
+            0.0
+          }
+        case None =>
+          sum += coefficients(j) * coefficients(j)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/optim/loss/DifferentiableRegularization.scala]: Not sure if this is performance critical but in a few blocks like this an array index is dereferenced many times and could be saved off, if it mattered, to optimize a bit

**Review by srowen — COMMENTED** [review]
Looks OK to me. This cuts down duplication and adds tests, and as I understand paves the way for some further improvements.

**Review by srowen — APPROVED** [review]
I reviewed this with Seth too and it looks pretty good to me.  CCing @MLnick @jkbradley @yanboliang @hhbyyh for final comments

**Comment by srowen** [comment]
Merging tomorrow if there are no objections.

**Comment by MLnick** [comment]
Overall looks good to me. I think it's a good step to clean up the codebase and reduce the duplicated code.

I think the impl is pretty well thought through. A few comments (that probably should be part of follow up):

1. I'd like to think about ways the regularizer(s) could be made more like "function composition" since the loss and reg are both just `DiffFunctions`
2. I think there should be scope for factoring out the loss functions into some sort of traits to make things cleaner

The one thing that doesn't feel quite right is the fact that the std scaling finds its way into `L2Regularization` - it sort of feels like the abstraction is leaking there. Not quite sure how to address it (perhaps we could look at something in line with the way OWLQN does it's L1 reg?).

**Comment by sethah** [comment]
@MLnick I completely agree about the leaky regularization abstraction. In fact, I think the function composition feature would make it easy to get rid of that problem. Consider:

In the standardized features case we want to compute dL/dBj_std where Bj_std = Bj / sigmaj. Define g(x) = sum(x_j^2) (l2 reg) and f(y) = y_j / sigma_j (standardization). Then we could do `val h = g.compose(f)` to give a loss function that provides the derivative of h as dg/df * df/dx. I hope that made some sense. Right now, it seems an over-engineered solution to me since we only use this function in a few places. It's definitely more elegant and more general, but I'd prefer to do that as a follow up or if we decide to keep building new algorithms using these abstractions.

Can you expand on your second point? Thanks!

**Comment by MLnick** [comment]
Sure, makes sense. We can always consider it later. Or even an alternate version of it to have `L2` and a subclass `StandardizedL2` or whatever (that's more if we were to start thinking about exposing the building blocks to external algorithm developers).

For point (2), it's just that each loss function ("squared loss", "logistic", etc) can implement a `Loss` trait similar to the old `org.apache.spark.mllib.optimization.Gradient` approach. The `Loss` would then be an arg of the `Aggregator` I suppose and the `add` method could be further consolidated. Not sure if it adds that much value here because of the funky standardization stuff we do in LiR and LoR...

**Comment by MLnick** [comment]
But even for the standard-scaling - it seems that could be expressed generically too with respect to scaling the coeff and gradient during the computation. Again, something perhaps for later.

**Comment by sethah** [comment]
Ok, yes all good points. I think since these are all private apis it gives us room for future changes. For now, I think we can get rid of a lot of code duplication and fill in some testing gaps with this change. I will be happy to drive the roadmap on this front as we move forward.

**Comment by srowen** [comment]
@sethah @MLnick am I reading right that this can be merged as a step forward?

**Comment by sethah** [comment]
@srowen Speaking for myself, I think the other concerns can be issued as follow ups, yes.

**Comment by srowen** [comment]
Merged to master

**Comment by MLnick** [comment]
Sorry for not replying - yeah agree let's create a set of follow ups to
this for potential improvements, and of course the migration of other
models to the framework

Thanks @sethah!

**Pull request closed without merging**